internal/bgp/frr/frr.go

```diff
@@ -287,96 +273,97 @@ func (sm *sessionManager) createConfig() (*frrConfig, error) {
+			frrFamily := frrIPFamily(family)
 
-			// Convert community 32bits value to : format
 			for _, c := range adv.Communities {
+				prefixListName := communityPrefixList(neighbor, c.String(), frrFamily)
 				if community.IsLarge(c) {
-					largeCommunities = append(largeCommunities, c.String())
-					if family == ipfamily.IPv4 {
-						properties.LargeCommunitiesV4.Insert(c.String())
-						continue
-					}
-					properties.LargeCommunitiesV6.Insert(c.String())
-					continue
+					prefixListName = largeCommunityPrefixList(neighbor, c.String(), frrFamily)
 				}
-				communities = append(communities, c.String())
-				if family == ipfamily.IPv4 {
-					properties.CommunitiesV4.Insert(c.String())
-					continue
+				prefixList, ok := neighbor.PrefixesModifiers[prefixListName]
+				if !ok {
+					prefixList = PropertyPrefixList{
+						Name:         prefixListName,
+						IPFamily:     frrFamily,
+						SetStatement: setStatementForCommunity(c),
+						prefixesSet:  sets.New[string](),
+						Prefixes:     []string{},
+					}
 				}
-				properties.CommunitiesV6.Insert(c.String())
-			}
-
-			prefix := adv.Prefix.String()
-			advConfig := advertisementConfig{
-				IPFamily:         family,
-				Prefix:           prefix,
-				Communities:      sort.StringSlice(communities),
-				LargeCommunities: sort.StringSlice(largeCommunities),
-				LocalPref:        adv.LocalPref,
+				prefixList.prefixesSet.Insert(prefix)
+				neighbor.PrefixesModifiers[prefixListName] = prefixList
 			}
 			if adv.LocalPref != 0 {
-				if family == ipfamily.IPv4 {
-					properties.LocalPrefsV4.Insert(adv.LocalPref)
-				} else {
-					properties.LocalPrefsV6.Insert(adv.LocalPref)
+				prefixListName := localPrefPrefixList(neighbor, adv.LocalPref, frrFamily)
+				prefixList, ok := neighbor.PrefixesModifiers[prefixListName]
+				if !ok {
+					prefixList = PropertyPrefixList{
+						Name:         prefixListName,
+						IPFamily:     frrFamily,
+						SetStatement: setStatementForLocalPref(adv.LocalPref),
+						prefixesSet:  sets.New[string](),
+						Prefixes:     []string{},
+					}
 				}
-			}
-
-			neighbor.Advertisements, err = addToAdvertisements(neighbor.Advertisements, &advConfig)
-			if err != nil {
-				return nil, err
+				prefixList.prefixesSet.Insert(prefix)
+				neighbor.PrefixesModifiers[prefixListName] = prefixList
 			}
 
 			switch family {
 			case ipfamily.IPv4:
+				neighbor.prefixesV4Set.Insert(prefix)
 				rout.ipV4Prefixes[prefix] = prefix
-				neighbor.HasV4Advertisements = true
 			case ipfamily.IPv6:
+				neighbor.prefixesV6Set.Insert(prefix)
 				rout.ipV6Prefixes[prefix] = prefix
-				neighbor.HasV6Advertisements = true
 			}
 		}
 	}
 
 	for _, r := range sortMap(routers) {
+		for _, n := range r.neighbors {
+			n.PrefixesV4 = sets.List(n.prefixesV4Set)
+			n.PrefixesV6 = sets.List(n.prefixesV6Set)
+			for k, m := range n.PrefixesModifiers {
+				m.Prefixes = sets.List(n.PrefixesModifiers[k].prefixesSet)
+				n.PrefixesModifiers[k] = m
+			}
```

Comment:
is there a reason not to have this above where we handle the neighbor? (also, feels like the prefixesV*Set can be dropped as a field if we do that)

Reply:
yep, it's because the current implementation runs under the assumption that we could have multiple sessions associated to the same neighbor. So it's better to sort once the negihbors and advertisements are squashed